### ISSUE TITLE
[CPU] Training is excruciatingly slow for CatBoostClassifier

### ISSUE THREAD

**Issue by JivanRoquet** [catboost/catboost#1859]
- Problem: (very) slow training
- catboost version: latest
- Operating System: macOS
- CPU: 2 GHz Quad-Core Intel Core i5

Using the Python client, when training on a dataset with 34k rows and 25 features (among which 22 are categorical), with a multiclass categorical target label of about 75 distinct classes, and using the following hyper parameters:

```python
model = CatBoostClassifier(
    n_estimators=60,
    learning_rate=0.1,
    depth=6,
    max_ctr_complexity=1,
    loss_function='MultiClass'
)
```

Training time takes about 15 minutes. If I raise `max_ctr_complexity` or `n_estimators` this can get to multiple hours.

Training with LightGBM on the exact same dataset with similar hyper parameters takes approx. 20 seconds, and gives similar predicting performance. A bit more or less depending on the exact parameters, but always in the same ballpark.

Am I doing something wrong?

**Comment by kizill** [comment]
Hello! First of all, there is a huge difference in default categorical features encoding in CatBoost & LightGBM, as we are trying to calculate 37 (number of classes) * 22 (number of cat features) * 3 (number of default counter priors) and then to score this much (~ 2442) additional feature columns (yes, everything is more complex, but i guess this is a good approximation). In contrary, as i know, LightGBM only tries 1 counter and one-hot encoding on top of single categorical feature. To speedup things, you may set `one_hot_max_size=254` to allow use one-hot encoding instead of counters for features with <=254 distinct values. The main idea here is to use counters only when they are really needed, for example for categorical features with high cardinality.
For more tips on speeding up training - try this https://catboost.ai/en/docs-beta/concepts/speed-up-training#one-hot-encoding 
And please, return with some news on how all this helped you :)

**Comment by kizill** [comment]
Ping 🐱

**Comment by JivanRoquet** [comment]
Hello @kizill and many thanks for the guidance.

I understand that CatBoost can be tuned for better speed, but it seems that training will still be orders of magnitude slower than LightGBM. This is fine, as for now we do all our training with LightGBM for faster iteration work, and once we're happy with our dataset and results, we might turn to CatBoost for production, as training time will be much less of an issue there.

I just would not recommend CatBoost as a tool for iterative discovery, as LightGBM offers much quicker feedback loop when experimenting with a given problem. CatBoost seems to provide better accuracy though, so definitely appropriate for production use.

I'm yet to try the `one_hot_max_size=254` bit though. It it does make training time similar to LightGBM I'd be happy to switch to CatBoost for exploratory iterative work as well. But if it only decreases training time by a factor of 2 or 3, this would still be 5x slower than LightGBM.

**Comment by kizill** [comment]
Can you help prepare a sample that will allow us to reproduce your setup? Random data with matching cat feature cardinality will work for sure:)

**Comment by kizill** [comment]
Closed as inactive, feel free to reopen.

**Comment by jmakov** [comment]
That's also my experience (and from the look of the issues at the pycaret project), also from many others. Is this expected?
